packages/mui-base/src/utils/useTransitionStatus.ts

```diff
@@ -0,0 +1,52 @@
+'use client';
+import * as React from 'react';
+import { useEnhancedEffect } from './useEnhancedEffect';
+
+export type TransitionStatus = 'entering' | 'exiting' | undefined;
+
+/**
+ * Provides a status string for CSS animations.
+ * @param open - a boolean that determines if the element is open.
+ * @ignore - internal hook.
+ */
+export function useTransitionStatus(open: boolean) {
+  const [transitionStatus, setTransitionStatus] = React.useState<TransitionStatus>();
+  const [mounted, setMounted] = React.useState(open);
+
+  if (open && !mounted) {
+    setMounted(true);
+  }
+
+  if (!open && mounted && transitionStatus !== 'exiting') {
+    setTransitionStatus('exiting');
+  }
+
+  if (!open && !mounted && transitionStatus === 'exiting') {
+    setTransitionStatus(undefined);
+  }
+
+  useEnhancedEffect(() => {
+    if (!open) {
+      return undefined;
+    }
+
+    setTransitionStatus('entering');
+
+    const frame = requestAnimationFrame(() => {
+      setTransitionStatus(undefined);
+    });
+
+    return () => {
+      cancelAnimationFrame(frame);
+    };
+  }, [open]);
+
+  return React.useMemo(
+    () => ({
+      mounted,
+      setMounted,
```

Comment:
Does it make sense to call `setMounted(true)` from outside of this hook? If not, we could return an `onAnimationEnd` callback that will call `setMounted(false)` (and maybe also `setTransitionStatus(undefined)`?)

Reply:
`setMounted` gives them the freedom to perform more complex checks and keeps its responsibility minimal. For example, due to the wrapper in anchored popups, the checks in `animationend`/`transitionend` may differ. This gives full control.